### PR TITLE
[Feature] implement left semi/anti join for NLJ

### LINKED ISSUE #10525: [sqlancer] query error
### Steps to reproduce the behavior (Required)
```
DROP DATABASE IF EXISTS w2;
CREATE DATABASE w2;
USE w2;
admin set frontend config ("enable_decimal_v3" = "true");
CREATE TABLE IF NOT EXISTS t0 (c_0_0 DECIMAL(34, 3)  NOT NULL ,c_0_1 DATETIME NOT NULL ,c_0_2 BOOLEAN NOT NULL ,c_0_3 BOOLEAN NULL ,c_0_4 DATE NULL ,c_0_5 DATETIME NOT NULL ,c_0_6 DECIMAL(22, 1)  NULL ,c_0_7 BOOLEAN NULL ,c_0_8 DATE NULL ,c_0_9 DATETIME NULL ,c_0_10 DECIMAL(35, 2)  NULL ,c_0_11 BOOLEAN NOT NULL ,c_0_12 VARCHAR(1) NOT NULL ,c_0_13 DECIMAL(16, 7)  NULL ,c_0_14 DECIMAL(34, 17)  NOT NULL ,c_0_15 BOOLEAN NULL ,c_0_16 DECIMAL(12, 2)  NOT NULL ,c_0_17 DATE NULL ,c_0_18 CHAR(1) NULL , INDEX index1_c_0_11(`c_0_11`) USING BITMAP) UNIQUE KEY (c_0_0,c_0_1,c_0_2,c_0_3,c_0_4,c_0_5,c_0_6,c_0_7,c_0_8,c_0_9,c_0_10,c_0_11,c_0_12,c_0_13,c_0_14,c_0_15,c_0_16) PARTITION BY RANGE(c_0_1) (START ("2010-01-01") END ("2022-12-31") EVERY (INTERVAL 115 day)) DISTRIBUTED BY HASH (c_0_13,c_0_3) BUCKETS 3 properties("replication_num"="3") ;
CREATE TABLE t1 (c_1_0 DECIMAL(36, 14)  NULL ,c_1_1 DATE NULL ,c_1_2 BOOLEAN NULL ,c_1_3 BOOLEAN NULL ,c_1_4 DECIMAL(1, 0)  NOT NULL ,c_1_5 DATETIME NULL ,c_1_6 BOOLEAN NOT NULL ,c_1_7 DOUBLE NOT NULL ,c_1_8 BIGINT NULL ,c_1_9 FLOAT NOT NULL ,c_1_10 DATETIME NULL ,c_1_11 DOUBLE NULL ,c_1_12 JSON NULL ,c_1_13 DECIMAL(2, 0)  NULL ,c_1_14 FLOAT NOT NULL , INDEX index1_c_1_3(`c_1_3`) USING BITMAP) DUPLICATE KEY (c_1_0,c_1_1,c_1_2,c_1_3,c_1_4,c_1_5,c_1_6) PARTITION BY RANGE(c_1_1) (START ("2010-01-01") END ("2022-12-31") EVERY (INTERVAL 153 day)) DISTRIBUTED BY HASH (c_1_2) BUCKETS 3 properties("replication_num"="3") ;
DELETE FROM t1 WHERE c_1_0=79994360.91786438939629;
SELECT t1_36.c_1_0 FROM t1 AS t1_36 WHERE (NOT (EXISTS (SELECT t0_35.c_0_0 FROM t0 AS t0_35 WHERE (t1_36.c_1_4) IN (3) ) )) ORDER BY t1_36.c_1_11 DESC;
--Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Required field 'eq_join_conjuncts' was not present! Struct: THashJoinNode(join_op:LEFT_ANTI_JOIN, eq_join_conjuncts:null, other_join_conjuncts:[TExpr(nodes:[TExprNode(node_type:BINARY_PRED, type:TTypeDesc(types:[TTypeNode(type:SCALAR, scalar_type:TScalarType(type:BOOLEAN))]), opcode:EQ, num_children:2, output_scale:-1, vector_opcode:INVALID_OPCODE, child_type:DECIMAL64, has_nullable_child:false, is_nullable:false, is_monotonic:false), TExprNode(node_type:SLOT_REF, type:TTypeDesc(types:[TTypeNode(type:SCALAR, scalar_type:TScalarType(type:DECIMAL64, precision:1, scale:0))]), num_children:0, slot_ref:TSlotRef(slot_id:5, tuple_id:0), output_scale:-1, output_column:-1, has_nullable_child:false, is_nullable:false, is_monotonic:true), TExprNode(node_type:DECIMAL_LITERAL, type:TTypeDesc(types:[TTypeNode(type:SCALAR, scalar_type:TScalarType(type:DECIMAL64, precision:1, scale:0))]), num_children:0, decimal_literal:TDecimalLiteral(value:3, integer_value:03 00 00 00 00 00 00 00), output_scale:-1, has_nullable_child:false, is_nullable:false, is_monotonic:true)])], is_push_down:false, sql_join_predicates:5: c_1_4 = 3, build_runtime_filters_from_planner:true, distribution_mode:BROADCAST, output_columns:[1, 12])

```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 - mysql> select current_version();
+----------------------+
| current_version()    |
+----------------------+
| MAIN-RELEASE aa27516 |
+----------------------+

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #10525

## Problem Summary(Required) ：

1. Implement `LEFT SEMI/ANTI` for `NestloopJoinOperator`.
2. Transform right semi/anti join to left semi/anti in optimizer

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-2.4

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 11 / 11 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/NestLoopJoinImplementationRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinCommutativityRule.java|8|8|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 11 / 11 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/NestLoopJoinImplementationRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinCommutativityRule.java|8|8|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 16 / 16 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/NestLoopJoinImplementationRule.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/NestLoopJoinNode.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinCommutativityRule.java|8|8|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

File: be/src/exec/pipeline/nljoin/nljoin_probe_operator.cpp [hunk 1]
```diff
@@ -271,21 +328,27 @@ ChunkPtr NLJoinProbeOperator::_permute_chunk(RuntimeState* state) {
     _probe_row_start = _probe_row_current;
     for (; _probe_row_current < _probe_chunk->num_rows(); ++_probe_row_current) {
         // Last build chunk must permute a chunk
-        if (_curr_build_chunk_index == _num_build_chunks() - 1 && _num_build_chunks() > 1) {
+        bool is_last_build_chunk = _curr_build_chunk_index == _num_build_chunks() - 1 && _num_build_chunks() > 1;
```
Comment (satanson) [inline be/src/exec/pipeline/nljoin/nljoin_probe_operator.cpp:331]: !_num_build_chunks() .empty() &&  _curr_build_chunk_index == _num_build_chunks() - 1 

when num_build_chunks().empty() return true, can this operator behavior correctly,  do you have add such test caces?  the empty build side must be result of the sub-plan to avoid a EmptySet node.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/implementation/NestLoopJoinImplementationRule.java [hunk 2]
```diff
@@ -48,12 +48,14 @@ public boolean check(final OptExpression input, OptimizerContext context) {
     }
 
     private boolean supportJoinType(JoinOperator joinType) {
-        return joinType.isCrossJoin() || joinType.isInnerJoin() || joinType.isOuterJoin();
+        return joinType.isCrossJoin() || joinType.isInnerJoin() || joinType.isOuterJoin() || joinType.isSemiAntiJoin();
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/implementation/NestLoopJoinImplementationRule.java:51]: We could only support the left anti/semi, but not right. Maybe keep this check is better.

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/NestLoopJoinTest.java [hunk 3]
```diff
@@ -115,8 +114,27 @@ public void testUnsupportedNLJoin() {
...
+        assertPlanContains("select * from t0,t1 where 1 not in (select v7 from t2,t3 where t0.v1 = 1 and t1.v4 = 2)",
+                "NESTLOOP JOIN");
+        assertPlanContains("select * from t0,t1 where v1 not in (select 1+2+3 from t2,t3 where t0.v1 = 1 and t1.v4 = 2)",
+                "NESTLOOP JOIN");
+        assertPlanContains("select * from t0,t1 where abs(1) - 1 not in (select v7 + 1 from t2,t3 where t0.v1 = 1 and t1.v4 = 2)",
+                "NESTLOOP JOIN");
+        assertPlanContains("select * from t0 left semi join t1 on t0.v1 < t1.v4", "NESTLOOP JOIN");
+        assertPlanContains("select * from t0 left anti join t1 on t0.v1 < t1.v4", "NESTLOOP JOIN");
+        assertPlanContains("select * from t0 right semi join t1 on t0.v1 < t1.v4", "NESTLOOP JOIN");
+        assertPlanContains("select * from t0 right anti join t1 on t0.v1 < t1.v4", "NESTLOOP JOIN");
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/NestLoopJoinTest.java:137]: It's the test case for semi/anti, other join types are tested in some other cases.

File: be/src/exec/pipeline/nljoin/nljoin_probe_operator.cpp (hunk 1 above)
Comment (mofeiatwork) [reply be/src/exec/pipeline/nljoin/nljoin_probe_operator.cpp:331]: Yes, empty build side is covered. But it is handled outside this function.

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run all

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 16 / 16 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/NestLoopJoinImplementationRule.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/NestLoopJoinNode.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinCommutativityRule.java|8|8|100.00%|[]|

**Review by satanson — DISMISSED** [review]

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 15 / 15 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/NestLoopJoinImplementationRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/NestLoopJoinNode.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinCommutativityRule.java|8|8|100.00%|[]|

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 15 / 15 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/NestLoopJoinImplementationRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/NestLoopJoinNode.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinCommutativityRule.java|8|8|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 16 / 16 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/NestLoopJoinImplementationRule.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/NestLoopJoinNode.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinCommutativityRule.java|8|8|100.00%|[]|
